src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -741,13 +741,32 @@ class NBDWatchCtx : public librbd::UpdateWatchCtx
       int ret;
 
       if (new_size != size) {
+        unsigned long _size = 0;
+
         dout(5) << "resize detected" << dendl;
         if (ioctl(fd, BLKFLSBUF, NULL) < 0)
           derr << "invalidate page cache failed: " << cpp_strerror(errno)
                << dendl;
-	if (use_netlink) {
+	if (!use_ioctl) {
 	  ret = netlink_resize(nbd_index, new_size);
-	} else {
+
+	  if (!ret) {
+            std::string nbd_path = "/sys/block/nbd" + stringify(nbd_index);
+            if(access(nbd_path.c_str(), F_OK) == 0) {
+              std::ifstream ifs;
+              ifs.open(nbd_path + "/size", std::ifstream::in);
+              if (ifs.is_open())
+                ifs >> _size;
+	    }
+	  }
+        }
+
+	/*
+	 * For old kernels maybe the netlink resize is not supported
+	 * yet, if so the kernel will do nothing and return success,
+	 * so let's fall back to use ioctl method.
+	 */
```

Comment:
nit: I would put this comment before you are checking the size, as it explains why we actually do this. And I think we could just set `use_ioctl` to true if "check device size" failed so we do not do this again on the next resize.